rclpy/rclpy/action_client.py

```diff
@@ -0,0 +1,495 @@
+# Copyright 2019 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+import time
+import uuid
+
+from action_msgs.msg import GoalStatus
+from action_msgs.srv import CancelGoal
+
+from rclpy.impl.implementation_singleton import rclpy_action_implementation as _rclpy_action
+# TODO(jacobperron): Move check_for_type_support to its own module (e.g. type_support)
+#                    Do after Crystal patch release since this breaks API
+from rclpy.node import check_for_type_support
+from rclpy.qos import qos_profile_action_status_default
+from rclpy.qos import qos_profile_default, qos_profile_services_default
+from rclpy.task import Future
+from rclpy.waitable import NumberOfEntities, Waitable
+
+from unique_identifier_msgs.msg import UUID
+
+
+class ClientGoalHandle():
+    """Goal handle for working with Action Clients."""
+
+    def __init__(self, goal_id, goal_response):
+        if not isinstance(goal_id, UUID):
+            raise TypeError('Expected UUID, but given {}'.format(type(goal_id)))
+
+        self._goal_id = goal_id
+        self._goal_response = goal_response
+        self._status = GoalStatus.STATUS_UNKNOWN
+
+    def __eq__(self, other):
+        return self._goal_id == other.goal_id
+
+    def __repr__(self):
+        return 'ClientGoalHandle <{}>'.format(self.goal_id.uuid)
+
+    @property
+    def goal_id(self):
+        return self._goal_id
+
+    @property
+    def goal_response(self):
+        return self._goal_response
+
+    @property
+    def accepted(self):
+        return self._goal_response.accepted
+
+    @property
+    def status(self):
+        return self._status
+
+
+class ActionClient(Waitable):
```

Comment:
This doesn't need to be done in this PR, or at all. I would recommend making `ActionClient` have a `Waitable` rather than inherit from it so `help(instance_of_action_client)` doesn't show the methods from the `Waitable` interface.

Reply:
Makes sense. What might be nicer is changing the `Waitable` API so that the methods have leading underscores. This should hide the `Waitable` methods and avoid having to create a subclass `ActionClientWaitable` for `ActionClient` to instantiate.